Add EC2 spaces test as assess_network_spaces.py

vernhart opened the pull request:
## Please provide the following details to expedite Pull Request review:

----

## Description of change

Add EC2 network spaces tests to assess_network_health.py

## QA steps

Part of QA.

## Documentation changes

Affects CI workflow.

nskaggs commented on acceptancetests/assess_network_health.py:
```diff
@@ -176,17 +184,214 @@ def connect_to_existing_model(self, client, target_model):
+            for subnet in subnets['subnets'].keys():
+                # skip subnets with INFAN in the provider-id as they may be
+                # inherited from underlay and therefore we cannot assign a space
+                if 'INFAN' not in subnets['subnets'][subnet]['provider-id']:
```
Maybe be more verbose here in our logging. Ensure we are logging if we skip or add a space and why.

nskaggs commented on acceptancetests/assess_network_health.py:
```diff
@@ -176,17 +184,214 @@ def connect_to_existing_model(self, client, target_model):
+            spaces = yaml.safe_load(client.get_juju_output('list-spaces',
+                '--format=yaml'))['spaces']
+            # Remove fan subnets from spaces list to avoid confusion
+            for space in list(spaces):
```
This probably is useful outside the method. Consider making a function for clarity and reuse.

Veebers commented on acceptancetests/assess_network_health.py:
```diff
@@ -325,6 +530,7 @@ def ensure_exposed(self, client, series):
             self.setup_expose_test(client, series, exposed)
 
         service_results = {}
+        log.info('MACHINE:\n {}'.format(json.dumps(yaml.safe_load(client.get_juju_output('list-machines', '--format=yaml'))['machines'], indent=4, sort_keys=True))) # VWH
```
This needs some linebreaks :-)

vernhart replied:
That line was meant for me debugging my code. The line itself is too long and the command gives excessive output. I'll break it up anyway. :)

Veebers commented on acceptancetests/assess_network_health.py:
```diff
@@ -502,6 +708,16 @@ def is_ipv6(self, address):
             return False
         return True
 
+    def ip_in_cidr(self, address, cidr):
```
This should be a standalone function, there is no need for it to be a member method.

nskaggs commented:
@vernhart , per @wupeka 

"we never create subnets (on any substrate), we need to have them precreated
To make it 'clean' I'd create a VPC and a couple of public and a couple of private subnets, then bootstrap into it and do what the test is doing"

vernhart commented on acceptancetests/assess_network_health.py:
```diff
@@ -13,6 +13,7 @@
 import os
 import socket
 from collections import defaultdict
+from netaddr import IPNetwork, IPAddress
```
I've replaced this with ipaddress, a built-in module in python3. I've added a line to requirements.txt for pre-3 versions. I'm not sure if python 3 resolves this requirement but I would expect it to.

vernhart commented on acceptancetests/assess_network_health.py:
```diff
@@ -176,17 +184,214 @@ def connect_to_existing_model(self, client, target_model):
+
+        :param client: Juju client object with controller
+        """
+        if client.env.provider == "ec2":
```
Where would be the best place to put supported_providers? As a global? In main and then passed as parameters to called functions? As self.supported_providers in the class __init__?
Does it make sense to rename this supported_spaces_providers to be explicit?

vernhart commented on acceptancetests/assess_network_health.py:
```diff
@@ -176,17 +184,214 @@ def connect_to_existing_model(self, client, target_model):
+            subnets = yaml.safe_load(client.get_juju_output('list-subnets',
+                '--format=yaml'))
+            if not subnets:
+                raise Exception('No subnets defined in EC2')
```
Would it make sense to create a custom exception (such as ConfigException) for cases like this? I browsed through the built-in exceptions and I'm not sure there is one that is perfectly applicable except maybe EnvironmentError: https://docs.python.org/2/library/exceptions.html#exceptions.EnvironmentError

vernhart commented on acceptancetests/assess_network_health.py:
```diff
@@ -176,17 +184,214 @@ def connect_to_existing_model(self, client, target_model):
+            machines = yaml.safe_load(client.get_juju_output('list-machines',
+                '--format=yaml'))['machines']
+            machine_test = self.verify_machine_spaces(client, spaces, machines)
+            log.info('Machines in Expected Spaces '
```
This makes sense. So, the verify functions should do their own logging and just return success/fail?

vernhart commented on acceptancetests/assess_network_health.py:
```diff
@@ -176,17 +184,214 @@ def connect_to_existing_model(self, client, target_model):
+            results[machine] = False
+            for ip in eth0['ip-addresses']:
+                if self.ip_in_cidr(ip, spaces[expected_space].keys()[0]):
+                    results[machine] = True
```
If one IP on eth0 is in the expected subnet, we're good. I'll add a break.

vernhart commented on acceptancetests/assess_network_health.py:
```diff
@@ -176,17 +184,214 @@ def connect_to_existing_model(self, client, target_model):
+                if self.ip_in_cidr(ip, spaces[expected_space].keys()[0]):
+                    results[machine] = True
+                else:
+                    log.info('In machine {0}, {1} is not in '
```
I had not encountered named parameters to the format method before. Thanks.

nskaggs commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Add/enhance spaces test](https://trello.com/c/RToJCqJv/66-add-enhance-spaces-test)

vernhart commented:
I've moved all of the spaces tests to assess_network_spaces.py and ensured they run without error. 
TODO:
- Set up VPC with proper subnetting before bootstrapping with existing client and clean up provisioning methods.
- Change assess methods to just return error strings instead of parsing the data later.
- Move supported_providers check logic to before setup.

vernhart commented:
@Veebers @wupeka @nskaggs 
This is ready to be reviewed (again), please.

Veebers commented on acceptancetests/assess_network_spaces.py:
```diff
@@ -0,0 +1,600 @@
+        failures = 0
+        failmsg = ''
+        alltests = [
+            self.verify_machine_spaces,
```
can we prefix these methods with test_ or similar to denote they are the testing calls.

vernhart replied:
I renamed a couple of the tests and prepended assert_.

Veebers commented on acceptancetests/assess_network_spaces.py:
```diff
@@ -0,0 +1,600 @@
+                expected_space = 'space{}'.format(machine)
+            eth0 = machines[machine]['network-interfaces']['eth0']
+            subnet = spaces['spaces'][expected_space].keys()[0]
+            for ip in eth0['ip-addresses']:
```
If I'm reading this right, than any IP in eth0['ip-addresses'] that isn't in the subnet cidr is a failure for the test? (and the successful ones don't really matter?)

This could be simplified (and IMO easier to read) as something like:
```python
ip_address = eth0['ip-addresses']
ips_in_correct_space = [ip for ip in ip_addresses if ip_in_cidr(ip, subnet)]
missing_ips = set(ip_addresses).difference(ips_in_correct_space)
if missing_ips != set():
    message = 'Machine {} IPs {} where not in ...'.format(', '.join(missing_ips))
```

vernhart replied:
It should be the other way around: As long as we have one IP on eth0 in the correct space, we're good. I see now that my logic fails if the good IP is not first.

Veebers replied:
Ah right, that simplifies it even further, it can be expressed as:

```python
if not any([ip for ip in ip_addresses if ip_in_cidr(ip, subnet)]):
    raise TestError('None of the IPs where in the CIDR . . .')
```

Veebers commented on acceptancetests/assess_network_spaces.py:
```diff
@@ -0,0 +1,600 @@
+        :param series: Ubuntu series to deploy
+        """
+        log.info("Adding 4 machines")
+        for x in range(0, 4):
```
The logic here can be removed by using:

```python
for space in [1, 1, 2, 3]:
```

vernhart replied:
Ah, good call.

Veebers commented on acceptancetests/assess_network_spaces.py:
```diff
@@ -0,0 +1,600 @@
+        ]
+
+        for test in alltests:
+            results = test(client)
```
The removal of needing a test to return a results with pass/fail/message will also shorten the test methods too improving their readability.

vernhart commented on acceptancetests/assess_network_spaces.py:
```diff
@@ -0,0 +1,600 @@
+
+
+
+class SubnetsNotReady(Exception):
```
Okay. Originally I was going to have the exception wrap the message with other text but decided against that.

vernhart commented on acceptancetests/assess_network_spaces.py:
```diff
@@ -0,0 +1,600 @@
+    """
+    if 'spaces' in subnets:
+        newsubnets['spaces'] = {}
+        for space in subnets['spaces'].keys():
```
Yes. Much more readable.

vernhart commented on acceptancetests/assess_network_spaces.py:
```diff
@@ -0,0 +1,600 @@
+    :param args: args from parse_args
+    """
+    supported_providers = [ 'ec2' ];
+    client = client_from_config(args.env,args.juju_bin)
```
Chicken and egg... We don't have a client object yet when the vpc is created. I'll test shuffling things around and see if it still works.

Veebers replied:
It should be fine, creating a BootstrapManager doesn't bootstrap anything, but it does have a client (bs_manager.client) that should have what's needed for this part.

vernhart replied:
👍

vernhart commented on acceptancetests/assess_network_spaces.py:
```diff
@@ -0,0 +1,600 @@
+    args = parse_args(argv)
+    configure_logging(args.verbose)
+
+    if args.model:
```
My understanding was that this was needed for the --existing functionality. I also assumed I was going to need this for bootstrapping into a pre-build vpc -- which turned out to not be true (and not work).

vernhart commented on acceptancetests/assess_network_spaces.py:
```diff
@@ -0,0 +1,600 @@
+
+class AssessNetworkSpaces:
+
+    def __init__(self, args):
```
I tried to remove all the cruft that I inherrited from copying the asssess_network_health.py script. I'll gladly remove more. :)

vernhart commented on acceptancetests/assess_network_spaces.py:
```diff
@@ -0,0 +1,600 @@
+            test.cleanup(client)
+            log.info('Cleanup complete.')
+
+def vpc_bootstrap(args):
```
I think I see where you're going here. You're saying we create a class, VPCAws that is a subclass of VPC so that we can add other subclasses later as spaces is implemented in other providers? I think the nomenclature is not quite right since VPC is an amazon thing. If we are to abstract this to other providers, the thing that is different is how spaces are set up.

Currently, two providers are supported, AWS (ec2) and MaaS. I don't know enough about how spaces under MaaS is different to know how much needs to be implemented differently.

Maybe instead of calling it VPC, we call it Spaces and we create subclasses SpacesAWS and SpacesMaaS. We'd have methods for pre_bootstrap and cleanup and maybe setup_spaces which would get called after we've bootstrapped but before running tests. Instead of exposing vpcid (which would not exist for MaaS), the pre_bootstrap method should take a bs_manager object as a parameter. The pre_bootstrap method would take care of feeding the vpcid to the bs_manager's client environment or setting up the MaaS excluded and reserved spaces.

vernhart commented:
Moved the pre_bootstrap function into a class for future expandability.
Corrected all the style problems reported by flake8.
And numerous stylistic simplifications as per @Veebers' recommendations.

Veebers commented:
LGTM
